File: shared/validation/validate.ts
/**
 * @fileoverview Validation utilities for Eigenvue step sequences and algorithm metadata.
 *
 * This module provides two layers of validation:
 *
 * 1. **Schema validation** — checks structural conformance against JSON Schema.
 *    Catches missing fields, wrong types, pattern mismatches, etc.
 *
 * 2. **Semantic validation** — checks invariants that JSON Schema cannot express.
 *    Catches index mismatches, terminal step violations, mathematical errors
 *    (e.g., attention weights not summing to 1.0), etc.
 *
 * Both layers return a ValidationResult with a boolean `valid` flag and an
 * array of human-readable `errors` strings.
 *
 * @module @eigenvue/validation
 */

import Ajv from "ajv/dist/2020";
import addFormats from "ajv-formats";

import stepFormatSchema from "../step-format.schema.json";
import metaSchema from "../meta.schema.json";
import { STEP_FORMAT_VERSION, type Step, type StepSequence } from "../types/step";

// ─────────────────────────────────────────────────────────────────────────────
// TYPES
// ─────────────────────────────────────────────────────────────────────────────

/** Result of a validation operation. */
export interface ValidationResult {
  /** True if the input is valid. */
  readonly valid: boolean;
  /**
   * Human-readable error messages. Empty if valid is true.
   * Each string is a self-contained error description.
   */
  readonly errors: readonly string[];
}

// ─────────────────────────────────────────────────────────────────────────────
// AJV SETUP (singleton — compiled once, reused)
// ─────────────────────────────────────────────────────────────────────────────

const ajv = new Ajv({
  allErrors: true,      // Report ALL errors, not just the first.
  verbose: true,        // Include schema and data in error objects.
  strict: true,         // Enforce strict schema authoring.
});
addFormats(ajv);        // Adds "date-time", "uri", etc.

const validateStepSequenceSchema = ajv.compile(stepFormatSchema);
const validateMetaSchema = ajv.compile(metaSchema);

// ─────────────────────────────────────────────────────────────────────────────
// STEP SEQUENCE VALIDATION
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Validate a step sequence against the JSON Schema AND semantic invariants.
 *
 * @param data - A parsed JSON object to validate as a StepSequence.
 * @returns A ValidationResult indicating success or listing all errors.
 *
 * @example
 * const result = validateStepSequence(parsedJson);
 * if (!result.valid) {
 *   console.error("Validation failed:", result.errors);
 * }
 */
export function validateStepSequence(data: unknown): ValidationResult {
  const errors: string[] = [];

  // ── Layer 1: JSON Schema validation ──

  const schemaValid = validateStepSequenceSchema(data);

  if (!schemaValid) {
    for (const err of validateStepSequenceSchema.errors ?? []) {
      errors.push(
        `Schema (${err.keyword}): ${err.instancePath || "/"} ${err.message ?? "unknown error"}` +
        (err.params ? ` (${JSON.stringify(err.params)})` : "")
      );
    }
    // If schema fails, semantic checks may crash — return early.
    return { valid: false, errors };
  }

  // ── Layer 2: Semantic validation (schema passed, so shape is correct) ──

  const seq = data as unknown as StepSequence;

  // 2a. Index contiguity: steps[i].index === i
  for (let i = 0; i < seq.steps.length; i++) {
    const step = seq.steps[i]!;
    if (step.index !== i) {
      errors.push(
        `Semantic: steps[${i}].index is ${step.index}, expected ${i}. ` +
        `Step indices must be contiguous and 0-based.`
      );
    }
  }

  // 2b. Terminal step: exactly one, and it must be the last
  const terminalIndices: number[] = [];
  for (let i = 0; i < seq.steps.length; i++) {
    if (seq.steps[i]!.isTerminal) {
      terminalIndices.push(i);
    }
  }

  if (terminalIndices.length === 0) {
    errors.push(
      `Semantic: No step has isTerminal=true. ` +
      `The last step must be terminal.`
    );
  } else if (terminalIndices.length > 1) {
    errors.push(
      `Semantic: Multiple steps have isTerminal=true at indices [${terminalIndices.join(", ")}]. ` +
      `Only the last step may be terminal.`
    );
  } else if (terminalIndices[0] !== seq.steps.length - 1) {
    errors.push(
      `Semantic: isTerminal=true is at index ${terminalIndices[0]}, ` +
      `but the last step is at index ${seq.steps.length - 1}. ` +
      `Only the last step may be terminal.`
    );
  }

  // 2c. Attention weights sum to 1.0 (±1e-6)
  for (let i = 0; i < seq.steps.length; i++) {
    const step = seq.steps[i]!;
    for (let j = 0; j < step.visualActions.length; j++) {
      const action = step.visualActions[j]!;
      if (action.type === "showAttentionWeights") {
        const weights = (action as { weights?: readonly number[] }).weights;
        if (Array.isArray(weights) && weights.length > 0) {
          const sum = weights.reduce((a: number, b: number) => a + b, 0);
          if (Math.abs(sum - 1.0) > 1e-6) {
            errors.push(
              `Semantic: steps[${i}].visualActions[${j}] (showAttentionWeights) ` +
              `weights sum to ${sum}, expected 1.0 (tolerance ±1e-6). ` +
              `This indicates a mathematical error in the softmax computation.`
            );
          }
          // Validate each weight is in [0, 1]
          for (let k = 0; k < weights.length; k++) {
            const w = weights[k]!;
            if (w < -1e-9 || w > 1.0 + 1e-9) {
              errors.push(
                `Semantic: steps[${i}].visualActions[${j}] (showAttentionWeights) ` +
                `weights[${k}] = ${w}, expected value in [0, 1].`
              );
            }
          }
        }
      }
    }
  }

  // 2d. HighlightRange: from <= to
  for (let i = 0; i < seq.steps.length; i++) {
    const step = seq.steps[i]!;
    for (let j = 0; j < step.visualActions.length; j++) {
      const action = step.visualActions[j]! as Record<string, unknown>;
      if (action.type === "highlightRange" || action.type === "dimRange") {
        const from = action.from as number | undefined;
        const to = action.to as number | undefined;
        if (from !== undefined && to !== undefined && from > to) {
          errors.push(
            `Semantic: steps[${i}].visualActions[${j}] (${action.type as string}) ` +
            `has from=${from} > to=${to}. 'from' must be <= 'to'.`
          );
        }
      }
    }
  }

  // 2e. CompareElements: valid result values
  for (let i = 0; i < seq.steps.length; i++) {
    const step = seq.steps[i]!;
    for (let j = 0; j < step.visualActions.length; j++) {
      const action = step.visualActions[j]! as Record<string, unknown>;
      if (action.type === "compareElements") {
        const result = action.result;
        if (result !== "less" && result !== "greater" && result !== "equal") {
          errors.push(
            `Semantic: steps[${i}].visualActions[${j}] (compareElements) ` +
            `result="${String(result)}", expected "less" | "greater" | "equal".`
          );
        }
      }
    }
  }

  // 2f. UpdateBarChart: labels length matches values length (if labels present)
  for (let i = 0; i < seq.steps.length; i++) {
    const step = seq.steps[i]!;
    for (let j = 0; j < step.visualActions.length; j++) {
      const action = step.visualActions[j]! as Record<string, unknown>;
      if (action.type === "updateBarChart") {
        const values = action.values as unknown[] | undefined;
        const labels = action.labels as unknown[] | undefined;
        if (values && labels && values.length !== labels.length) {
          errors.push(
            `Semantic: steps[${i}].visualActions[${j}] (updateBarChart) ` +
            `has ${values.length} values but ${labels.length} labels. ` +
            `If labels are provided, their count must match values.`
          );
        }
      }
    }
  }

  return { valid: errors.length === 0, errors };
}

// ─────────────────────────────────────────────────────────────────────────────
// ALGORITHM METADATA VALIDATION
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Validate an algorithm's meta.json against the JSON Schema AND semantic invariants.
 *
 * @param data - A parsed JSON object to validate as AlgorithmMeta.
 * @returns A ValidationResult indicating success or listing all errors.
 */
export function validateAlgorithmMeta(data: unknown): ValidationResult {
  const errors: string[] = [];

  // ── Layer 1: JSON Schema ──

  const schemaValid = validateMetaSchema(data);

  if (!schemaValid) {
    for (const err of validateMetaSchema.errors ?? []) {
      errors.push(
        `Schema (${err.keyword}): ${err.instancePath || "/"} ${err.message ?? "unknown error"}` +
        (err.params ? ` (${JSON.stringify(err.params)})` : "")
      );
    }
    return { valid: false, errors };
  }

  // ── Layer 2: Semantic validation ──

  const meta = data as Record<string, unknown>;

  // 2a. Quiz correctIndex within bounds
  const education = meta.education as Record<string, unknown>;
  const quiz = education.quiz as Array<Record<string, unknown>>;
  for (let i = 0; i < quiz.length; i++) {
    const q = quiz[i]!;
    const options = q.options as string[];
    const correctIndex = q.correctIndex as number;
    if (correctIndex < 0 || correctIndex >= options.length) {
      errors.push(
        `Semantic: education.quiz[${i}].correctIndex is ${correctIndex}, ` +
        `but options has ${options.length} items. Must be in [0, ${options.length - 1}].`
      );
    }
  }

  // 2b. defaultLanguage must exist in implementations
  const code = meta.code as Record<string, unknown>;
  const implementations = code.implementations as Record<string, unknown>;
  const defaultLanguage = code.defaultLanguage as string;
  if (!(defaultLanguage in implementations)) {
    errors.push(
      `Semantic: code.defaultLanguage is "${defaultLanguage}" ` +
      `but it is not a key in code.implementations. ` +
      `Available: [${Object.keys(implementations).join(", ")}].`
    );
  }

  return { valid: errors.length === 0, errors };
}

File: shared/validation/validate.py
"""
Eigenvue Validation Utilities — Python

Provides two layers of validation for step sequences and algorithm metadata:

1. Schema validation — structural conformance against JSON Schema.
2. Semantic validation — invariants that JSON Schema cannot express.

Usage:
    from shared.validation.validate import validate_step_sequence, validate_meta

    result = validate_step_sequence(parsed_json)
    if not result.valid:
        for error in result.errors:
            print(error)
"""

from __future__ import annotations

import json
import math
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any

import jsonschema
from jsonschema import Draft202012Validator, ValidationError

# ─────────────────────────────────────────────────────────────────────────────
# CONSTANTS
# ─────────────────────────────────────────────────────────────────────────────

_SCHEMA_DIR = Path(__file__).parent.parent
_STEP_FORMAT_SCHEMA_PATH = _SCHEMA_DIR / "step-format.schema.json"
_META_SCHEMA_PATH = _SCHEMA_DIR / "meta.schema.json"

# ─────────────────────────────────────────────────────────────────────────────
# SCHEMA LOADING (loaded once at module import time)
# ─────────────────────────────────────────────────────────────────────────────

def _load_schema(path: Path) -> dict[str, Any]:
    """Load and parse a JSON Schema file.

    Args:
        path: Absolute or relative path to the .schema.json file.

    Returns:
        Parsed JSON Schema as a dict.

    Raises:
        FileNotFoundError: If the schema file does not exist.
        json.JSONDecodeError: If the file is not valid JSON.
    """
    with open(path, "r", encoding="utf-8") as f:
        return json.load(f)


_STEP_FORMAT_SCHEMA: dict[str, Any] = _load_schema(_STEP_FORMAT_SCHEMA_PATH)
_META_SCHEMA: dict[str, Any] = _load_schema(_META_SCHEMA_PATH)

# Pre-compile validators for performance (schema is checked once, reused).
_step_format_validator = Draft202012Validator(_STEP_FORMAT_SCHEMA, format_checker=jsonschema.FormatChecker())
_meta_validator = Draft202012Validator(_META_SCHEMA, format_checker=jsonschema.FormatChecker())

# ─────────────────────────────────────────────────────────────────────────────
# RESULT TYPE
# ─────────────────────────────────────────────────────────────────────────────

@dataclass(frozen=True, slots=True)
class ValidationResult:
    """Result of a validation operation.

    Attributes:
        valid: True if the input passed all checks.
        errors: List of human-readable error messages. Empty if valid is True.
    """
    valid: bool
    errors: tuple[str, ...] = field(default_factory=tuple)


# ─────────────────────────────────────────────────────────────────────────────
# STEP SEQUENCE VALIDATION
# ─────────────────────────────────────────────────────────────────────────────

# Floating-point tolerance for attention weight sums and individual weights.
_FLOAT_TOLERANCE = 1e-6


def validate_step_sequence(data: Any) -> ValidationResult:
    """Validate a step sequence against JSON Schema and semantic invariants.

    Checks performed:
    - JSON Schema structural validation (fields, types, patterns).
    - Index contiguity: steps[i]["index"] == i for all i.
    - Terminal step: exactly one, and it is the last step.
    - Attention weights sum to 1.0 (±1e-6) and each weight is in [0, 1].
    - HighlightRange/DimRange: from <= to.
    - CompareElements result is one of "less", "greater", "equal".
    - UpdateBarChart labels length matches values length.

    Args:
        data: A parsed JSON object (dict) to validate.

    Returns:
        ValidationResult with valid=True if all checks pass,
        or valid=False with a list of all errors found.
    """
    errors: list[str] = []

    # ── Layer 1: JSON Schema ──

    schema_errors = list(_step_format_validator.iter_errors(data))
    if schema_errors:
        for err in schema_errors:
            path = "/".join(str(p) for p in err.absolute_path) or "/"
            errors.append(f"Schema ({err.validator}): /{path} {err.message}")
        # Schema failures may make semantic checks unsafe — return early.
        return ValidationResult(valid=False, errors=tuple(errors))

    # ── Layer 2: Semantic validation ──

    steps: list[dict[str, Any]] = data["steps"]

    # 2a. Index contiguity
    for i, step in enumerate(steps):
        if step["index"] != i:
            errors.append(
                f"Semantic: steps[{i}].index is {step['index']}, expected {i}. "
                f"Step indices must be contiguous and 0-based."
            )

    # 2b. Terminal step — exactly one, at the end
    terminal_indices = [i for i, s in enumerate(steps) if s["isTerminal"]]

    if len(terminal_indices) == 0:
        errors.append(
            "Semantic: No step has isTerminal=true. The last step must be terminal."
        )
    elif len(terminal_indices) > 1:
        errors.append(
            f"Semantic: Multiple steps have isTerminal=true at indices "
            f"{terminal_indices}. Only the last step may be terminal."
        )
    elif terminal_indices[0] != len(steps) - 1:
        errors.append(
            f"Semantic: isTerminal=true is at index {terminal_indices[0]}, "
            f"but the last step is at index {len(steps) - 1}. "
            f"Only the last step may be terminal."
        )

    # 2c. Attention weights: sum to 1.0 ±1e-6, each in [0, 1]
    for i, step in enumerate(steps):
        for j, action in enumerate(step["visualActions"]):
            if action["type"] == "showAttentionWeights":
                weights = action.get("weights", [])
                if isinstance(weights, list) and len(weights) > 0:
                    weight_sum = math.fsum(weights)  # fsum for maximum precision
                    if abs(weight_sum - 1.0) > _FLOAT_TOLERANCE:
                        errors.append(
                            f"Semantic: steps[{i}].visualActions[{j}] "
                            f"(showAttentionWeights) weights sum to {weight_sum}, "
                            f"expected 1.0 (tolerance ±{_FLOAT_TOLERANCE}). "
                            f"This indicates a mathematical error in the softmax computation."
                        )
                    for k, w in enumerate(weights):
                        if w < -_FLOAT_TOLERANCE or w > 1.0 + _FLOAT_TOLERANCE:
                            errors.append(
                                f"Semantic: steps[{i}].visualActions[{j}] "
                                f"(showAttentionWeights) weights[{k}] = {w}, "
                                f"expected value in [0, 1]."
                            )

    # 2d. HighlightRange / DimRange: from <= to
    for i, step in enumerate(steps):
        for j, action in enumerate(step["visualActions"]):
            if action["type"] in ("highlightRange", "dimRange"):
                from_val = action.get("from")
                to_val = action.get("to")
                if (
                    from_val is not None
                    and to_val is not None
                    and from_val > to_val
                ):
                    errors.append(
                        f"Semantic: steps[{i}].visualActions[{j}] ({action['type']}) "
                        f"has from={from_val} > to={to_val}. 'from' must be <= 'to'."
                    )

    # 2e. CompareElements result
    for i, step in enumerate(steps):
        for j, action in enumerate(step["visualActions"]):
            if action["type"] == "compareElements":
                result = action.get("result")
                if result not in ("less", "greater", "equal"):
                    errors.append(
                        f"Semantic: steps[{i}].visualActions[{j}] (compareElements) "
                        f'result="{result}", expected "less" | "greater" | "equal".'
                    )

    # 2f. UpdateBarChart labels/values length match
    for i, step in enumerate(steps):
        for j, action in enumerate(step["visualActions"]):
            if action["type"] == "updateBarChart":
                values = action.get("values", [])
                labels = action.get("labels")
                if labels is not None and len(values) != len(labels):
                    errors.append(
                        f"Semantic: steps[{i}].visualActions[{j}] (updateBarChart) "
                        f"has {len(values)} values but {len(labels)} labels. "
                        f"If labels are provided, their count must match values."
                    )

    return ValidationResult(valid=len(errors) == 0, errors=tuple(errors))


# ─────────────────────────────────────────────────────────────────────────────
# ALGORITHM METADATA VALIDATION
# ─────────────────────────────────────────────────────────────────────────────

def validate_meta(data: Any) -> ValidationResult:
    """Validate an algorithm's meta.json against JSON Schema and semantic invariants.

    Checks performed:
    - JSON Schema structural validation.
    - Quiz correctIndex within options bounds.
    - defaultLanguage exists in implementations.

    Args:
        data: A parsed JSON object (dict) to validate.

    Returns:
        ValidationResult with valid=True if all checks pass.
    """
    errors: list[str] = []

    # ── Layer 1: JSON Schema ──

    schema_errors = list(_meta_validator.iter_errors(data))
    if schema_errors:
        for err in schema_errors:
            path = "/".join(str(p) for p in err.absolute_path) or "/"
            errors.append(f"Schema ({err.validator}): /{path} {err.message}")
        return ValidationResult(valid=False, errors=tuple(errors))

    # ── Layer 2: Semantic validation ──

    # 2a. Quiz correctIndex within bounds
    for i, q in enumerate(data["education"]["quiz"]):
        options = q["options"]
        correct_index = q["correctIndex"]
        if correct_index < 0 or correct_index >= len(options):
            errors.append(
                f"Semantic: education.quiz[{i}].correctIndex is {correct_index}, "
                f"but options has {len(options)} items. "
                f"Must be in [0, {len(options) - 1}]."
            )

    # 2b. defaultLanguage must be a key in implementations
    default_lang = data["code"]["defaultLanguage"]
    implementations = data["code"]["implementations"]
    if default_lang not in implementations:
        errors.append(
            f'Semantic: code.defaultLanguage is "{default_lang}" '
            f"but it is not a key in code.implementations. "
            f"Available: [{', '.join(implementations.keys())}]."
        )

    return ValidationResult(valid=len(errors) == 0, errors=tuple(errors))
